src/main/scala/firrtl/passes/RemoveAccesses.scala

```diff
@@ -0,0 +1,166 @@
+    * E.g. if called on a[i] where a: UInt[2], we would return:
+    *   Seq(Location(a[0], UIntLiteral(0)), Location(a[1], UIntLiteral(1)))
+    */
+  def getLocations(e: Expression): Seq[Location] = e match {
+    case e: WRef => create_exps(e).map(Location(_,one))
+    case e: WSubIndex =>
+      val ls = getLocations(e.exp)
+      val start = get_point(e)
+      val end = start + get_size(tpe(e))
+      val stride = get_size(tpe(e.exp))
+      val lsx = mutable.ArrayBuffer[Location]()
+      var c = 0
+      for (i <- 0 until ls.size) {
+        if (((i % stride) >= start) & ((i % stride) < end)) {
+          lsx += ls(i)
+        }
+      }
+      lsx
+    case e: WSubField =>
+      val ls = getLocations(e.exp)
+      val start = get_point(e)
+      val end = start + get_size(tpe(e))
+      val stride = get_size(tpe(e.exp))
+      val lsx = mutable.ArrayBuffer[Location]()
+      var c = 0
+      for (i <- 0 until ls.size) {
+        if (((i % stride) >= start) & ((i % stride) < end)) { lsx += ls(i) }
+      }
+      lsx
+    case e: WSubAccess =>
+      val ls = getLocations(e.exp)
+      val stride = get_size(tpe(e))
+      val wrap = tpe(e.exp).asInstanceOf[VectorType].size
+      val lsx = mutable.ArrayBuffer[Location]()
+      var c = 0
+      for (i <- 0 until ls.size) {
+        if ((c % wrap) == 0) { c = 0 }
+        val basex = ls(i).base
+        val guardx = AND(ls(i).guard,EQV(uint(c),e.index))
+        lsx += Location(basex,guardx)
+        if ((i + 1) % stride == 0) {
+          c = c + 1
+        }
+      }
+      lsx
+  }
+  /** Returns true if e contains a [[firrtl.WSubAccess]]
+    */
+  def hasAccess(e: Expression): Boolean = {
+    var ret: Boolean = false
+    def rec_has_access(e: Expression): Expression = e match {
+      case (e:WSubAccess) => { ret = true; e }
+      case (e) => e map (rec_has_access)
+    }
+    rec_has_access(e)
+    ret
+  }
+  def run(c: Circuit): Circuit = {
+    def remove_m(m: Module): Module = {
+      val namespace = Namespace(m)
+      def onStmt(s: Statement): Statement = {
+        val stmts = mutable.ArrayBuffer[Statement]()
+        def create_temp(e: Expression): Expression = {
+          val n = namespace.newTemp
+          stmts += DefWire(info(s), n, tpe(e))
+          WRef(n, tpe(e), kind(e), gender(e))
+        }
+
+        /** Replaces a subaccess in a given male expression
+          */
+        def removeMale(e: Expression): Expression = e match {
+          case (_:WSubAccess| _: WSubField| _: WSubIndex| _: WRef) if (hasAccess(e)) => 
+            val rs = getLocations(e)
+            val foo = rs.find(x => {x.guard != one})
+            foo match {
+              case None => error("Shouldn't be here")
+              case foo: Some[Location] =>
+                val temp = create_temp(e)
+                val temps = create_exps(temp)
+                def getTemp(i: Int) = temps(i % temps.size)
+                (rs,0 until rs.size).zipped.foreach { (x,i) => 
```

Comment:
this is probably more clearly expressed with

    for ((x, i) <- rs.zipWithIndex) {